wordpress-mobile/WordPress-iOS
# Edit async images

Refs #13048 

Depends on https://github.com/wordpress-mobile/WordPress-iOS/pull/13069

Allows the user to edit "async images", ie.: images that need to be downloaded to be manipulated.

Keep in mind that this PR **does not cover** the integration with Gutenberg.

<img src="https://user-images.githubusercontent.com/7040243/70747348-8b29c080-1d06-11ea-85eb-251a9e2d6d35.gif" width="300" />

## To test

First, [apply this patch](https://gist.github.com/leandroalonso/131eb194eb38d7c6ea0f2eb1d7574f50).

*Creating a new post*

1. Create a new post in Gutenberg
2. Add an image
3. Tap the image twice
4. Check that the full quality image will be downloaded and then the crop capabilities are shown

*Editing an image added from web*

1. Using the browser, create a new post using Gutenberg
2. Add an image, then change "Image Size" to"Thumbnail" (in the sidebar at the right)
3. Save or publish the post
4. Open it on the app
5. Double-tap the image
6. Checks that the full quality image is downloaded and can be cropped/resized — not the thumbnail

*iPhone, iPad, Portrait and Landscape*

1. Please, test it using iPhone or iPad. Portrait and landscape!

## More information

To handle editing and manipulating "async images" an `AsyncImage` protocol was created. By conforming to that, every entity can be an async asset. :)

Let's get our app as an example: we have `PHAsset`, Aztec's attachment, Gutenberg Media... all of that can be given to Media Editor — the only requirement is to conform to the protocol.

You can see an example at the given patch above. An extension for `PHAsset` will be added soon. The protocol is quite simple, it contains:

* A `thumb` optional UIImage — if given, this will be displayed right away (as we did in the Gutenberg example);
* A `thumbnail` method. If no `thumb` was given, this method will be called. It is unlikely that we won't have a thumbnail but who knows...
* A `full` method. This method will be called to start fetching the image, and once is done to call the callback.
* A `cancel` method. This is called when the user gave up on editing the media while the asset is being downloaded. Here we should cancel all the ongoing operations (let's save some network data)

PR submission checklist:
